bluemira/equilibria/optimisation/constraints.py

```diff
@@ -445,6 +447,69 @@ def f_constraint(self) -> ConstraintFunction:
         return f_constraint
 
 
+class SphericalHarmonicsConstraint(MagneticConstraint):
+    """
+    Constraint function to constrain spherical harmonics starting from initial
+    coil currents and associated core plasma.
+
+    Parameters
+    ----------
+    r_t:
+        Typical length scale of the problem (e.g. radius at outer midplane)
+    ref_harmonics:
+        Initial harmonic amplitudes obtained from desired core plasma
+    max_degree:
+        Maximum degree of spherical harmonics desired to constrain.
+    """
+
+    def __init__(
+        self,
+        r_t: float,
+        ref_harmonics: npt.NDArray,
+        max_degree: int,
+        target_value: float = 0.0,
+        weights: Union[float, npt.NDArray] = 1.0,
+        tolerance: Union[float, npt.NDArray] = 1e-6,
+        f_constraint: Type[ConstraintFunction] = AxBConstraint,
+        constraint_type: str = "inequality",
+    ):
+        super().__init__(
+            target_value,
+            weights,
+            tolerance=tolerance,
+            f_constraint=f_constraint,
+            constraint_type=constraint_type,
+        )
+        self.max_degree = max_degree
+        self.r_t = r_t
+        self.ref_harmonics = ref_harmonics
+
+    def control_response(self, coilset: CoilSet) -> npt.NDArray:
+        """
+        Calculate control response of a CoilSet to the constraint.
+        """
+        return coil_harmonic_amplitude_matrix(coilset, self.max_degree, self.r_t)
+
+    def evaluate(self, _: Equilibrium) -> npt.NDArray:
+        """
+        Calculate the value of the constraint in an Equilibrium.
+        """
+        return self.ref_harmonics
+
+    def __len__(self) -> int:
+        """
+        The mathematical size of the constraint.
+        """
+        return len(self.ref_harmonics)
+
+    def plot(self, ax):
+        """
+        Plot the constraint onto an Axes.
+        """
+        # Possibly the r_t semi circle?
+        raise NotImplementedError("TODO")
```

Comment:
The only thing I could think of that would be possibly useful for this would be if you could plot the collocations points with their ref_amplitudes on the grid and maybe the coilset coil positions

Reply:
Just chatted with Georgi, it would be useful to plot the semi-circle (i.e. r_t) and the collocation points

Reply:
In its currrent form we don't have any information on the collocation points. Thats comes in before the max_degree is specified IIRC. For that information to be available I think the harmonic amplitudes and the max degree would need to be calculated in the constraint init. 
If thats a good direction and whats wanted it can be an improvement PR, probably outside the scope of this straight conversion.

I've added the r_t semi-circle